// @ngx-i18nsupport/tooling/schematics/src/common/common-testing_spec.js
"use strict";
/**
 * Some common functions used by testcases.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.workspaceOptions = {
    name: 'workspace',
    newProjectRoot: 'projects',
    version: '6.0.0',
};
function readAsJson(tree, path) {
    const content = tree.read(path);
    if (!content) {
        throw new Error('file ' + path + ' not found in tree');
    }
    const contentString = content.toString('UTF-8');
    return JSON.parse(contentString);
}
exports.readAsJson = readAsJson;
function readPackageJson(tree) {
    expect(tree.files).toContain('/package.json');
    return readAsJson(tree, '/package.json');
}
exports.readPackageJson = readPackageJson;
function readAngularJson(tree) {
    expect(tree.files).toContain('/angular.json');
    return readAsJson(tree, '/angular.json');
}
exports.readAngularJson = readAngularJson;
function readXliffmergeJson(projectName, tree) {
    const path = projectName ? `/projects/${projectName}/xliffmerge.json` : '/xliffmerge.json';
    expect(tree.files).toContain(path);
    return readAsJson(tree, path);
}
exports.readXliffmergeJson = readXliffmergeJson;
