// @ngx-i18nsupport/tooling/schematics/src/common/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
var constants_1 = require("./constants");
exports.xliffmergeVersion = constants_1.xliffmergeVersion;
exports.xliffmergePackage = constants_1.xliffmergePackage;
exports.defaultI18nFormat = constants_1.defaultI18nFormat;
exports.defaultI18nLocale = constants_1.defaultI18nLocale;
exports.extractScriptName = constants_1.extractScriptName;
var common_functions_1 = require("./common-functions");
exports.isValidLanguageSyntax = common_functions_1.isValidLanguageSyntax;
exports.addLanguageConfigurationToProject = common_functions_1.addLanguageConfigurationToProject;
exports.addStartScriptToPackageJson = common_functions_1.addStartScriptToPackageJson;
exports.buildConfigurationForLanguage = common_functions_1.buildConfigurationForLanguage;
exports.serveConfigurationForLanguage = common_functions_1.serveConfigurationForLanguage;
exports.fullExtractScript = common_functions_1.fullExtractScript;
tslib_1.__exportStar(require("./options-after-setup"), exports);
