// @ngx-i18nsupport/tooling/schematics/schematics-core/utility/special-project.js
"use strict";
/**
 * Additional angular.json spefific tool functions that are not part of normal project.ts
 */
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
/*
private helper function
 */
function readAngularJson(host) {
    const noAngularJsonMsg = 'file angular.json not found';
    if (host.exists('angular.json')) {
        const content = host.read('angular.json');
        if (!content) {
            throw new schematics_1.SchematicsException(noAngularJsonMsg);
        }
        const sourceText = content.toString('utf-8');
        return JSON.parse(sourceText);
    }
    else {
        throw new schematics_1.SchematicsException(noAngularJsonMsg);
    }
}
/**
 * Get project from angular.json.
 * @param host Host (Tree)
 * @param _context Context
 * @param projectName Name of project
 */
function getProjectByName(host, _context, projectName) {
    const json = readAngularJson(host);
    const projects = json['projects'];
    if (!projects) {
        throw new schematics_1.SchematicsException('angular.json does not contain projects');
    }
    const project = projects[projectName];
    if (!project) {
        throw new schematics_1.SchematicsException('angular.json does not contain project ' + projectName);
    }
    return project;
}
exports.getProjectByName = getProjectByName;
/**
 * (private) Add a configuration to angular.json.
 * Configuration is stored under path given as parameter.
 * @param host Host (Tree)
 * @param context Context
 * @param projectName Name of project
 * @param path path in project, like ['architect', 'build', 'configurations']
 * @param configurationName Name of configuration to add
 * @param configuration configuration object
 */
function addConfigurationToProjectPath(host, context, projectName, path, configurationName, configuration) {
    const json = readAngularJson(host);
    const projects = json['projects'];
    if (!projects) {
        throw new schematics_1.SchematicsException('angular.json does not contain projects');
    }
    const project = projects[projectName];
    if (!project) {
        throw new schematics_1.SchematicsException('angular.json does not contain project ' + projectName);
    }
    const configurations = getObjectFromProjectUsingPath(projectName, project, path);
    configurations[configurationName] = configuration;
    context.logger.debug('added configuration ' + configurationName + ' to project ' + projectName);
    host.overwrite('angular.json', JSON.stringify(json, null, 2));
    return host;
}
/**
 * Add a build configuration to angular.json.
 * Configuration is stored under architect.build.configurations
 * @param host Host (Tree)
 * @param context Context
 * @param projectName Name of project
 * @param configurationName Name of configuration to add
 * @param configuration configuration object
 */
function addArchitectBuildConfigurationToProject(host, context, projectName, configurationName, configuration) {
    return addConfigurationToProjectPath(host, context, projectName, ['architect', 'build', 'configurations'], configurationName, configuration);
}
exports.addArchitectBuildConfigurationToProject = addArchitectBuildConfigurationToProject;
/**
 * Add a serve configuration to angular.json.
 * Configuration is stored under architect.serve.configurations
 * @param host Host (Tree)
 * @param context Context
 * @param projectName Name of project
 * @param configurationName Name of configuration to add
 * @param configuration configuration object
 */
function addArchitectServeConfigurationToProject(host, context, projectName, configurationName, configuration) {
    return addConfigurationToProjectPath(host, context, projectName, ['architect', 'serve', 'configurations'], configurationName, configuration);
}
exports.addArchitectServeConfigurationToProject = addArchitectServeConfigurationToProject;
/**
 * (private) get a special object from project by navigating a path
 * Throws an exception if path does not exist.
 * @param projectName Name of project.
 * @param project the project read from angular.json
 * @param path path like ['architect', 'build', 'configurations']
 * @return the object at the path position
 */
function getObjectFromProjectUsingPath(projectName, project, path) {
    let object = project;
    let currentPath = '';
    for (let i = 0; i < path.length; i++) {
        currentPath = currentPath + '.' + path[i];
        object = object[path[i]];
        if (!object) {
            throw new schematics_1.SchematicsException('angular.json does not contain ' + currentPath + ' in project ' + projectName);
        }
    }
    return object;
}
/**
 * Read angular.json
 * @return content or null, if file does not exist.
 */
function getAngularJson(host) {
    const path = `/angular.json`;
    const content = host.read(path);
    if (!content) {
        return null;
    }
    const contentString = content.toString('UTF-8');
    return JSON.parse(contentString);
}
exports.getAngularJson = getAngularJson;
